go/vt/vtgate/engine/fake_primitive_test.go

```diff
@@ -30,34 +36,39 @@ type fakePrimitive struct {
 	curResult int
 	// sendErr is sent at the end of the stream if it's set.
 	sendErr error
+
+	log []string
 }
 
-func (tp *fakePrimitive) rewind() {
-	tp.curResult = 0
+func (f *fakePrimitive) rewind() {
+	f.curResult = 0
+	f.log = nil
 }
 
-func (tp *fakePrimitive) Execute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
-	if tp.results == nil {
-		return nil, tp.sendErr
+func (f *fakePrimitive) Execute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	f.log = append(f.log, fmt.Sprintf("Execute %v %v", printBindVars(bindVars), wantfields))
+	if f.results == nil {
+		return nil, f.sendErr
 	}
 
-	r := tp.results[tp.curResult]
-	tp.curResult++
+	r := f.results[f.curResult]
```

Comment:
One thing I noticed while writing tests for scattered limit offset feature, is that is easy to shoot in the foot with the results array here. Because it returns a pointer, you could end up in situations where the tests become tautological and you are comparing `x == x`. 

Would it make sense to return a copy of `r` here to avoid these kind of errors?

Reply:
If you look at the tests, these are usually for driving behavior, and not for verifying results.

Even in the case of verifying results, the comparison still makes sense because the result is often massaged by the primitives as they get returned through the call stack. In the case where it does not get modified, we still have to verify the correctness of the data.

Finally, I think the `DeepEqual` comparison essentially makes it equivalent to comparing copies, which should address your concern of tautology?

Let me know if I misunderstood your point.

Reply:
I think the way the tests are right now we are ok. But if you look [here](https://github.com/sougou/vitess/blob/f2951d28dc8215dcbf66686fc815670b5f26d41f/go/vt/vtgate/engine/limit_test.go#L69), `result` and `inputResult` are the same pointer (which is not obvious), so that test was never going to fail.